### DISCUSSION TITLE
Massive Speedup with Tiny AutoEncoder (Edit: No. It was a mistake)

### DISCUSSION THREAD

**Discussion by muerrilla** [PixArt-alpha/PixArt-alpha#31] [Ideas]
Just a heads up: 
If using diffusers, you can get a massive speedup by using [Tiny AutoEncoder](https://huggingface.co/docs/diffusers/api/models/autoencoder_tiny) instead of the original VAE (40% faster on my 4060ti) for a minor trade-off in quality.

EDIT: This was a mistake on my part. See [my comment](https://github.com/orgs/PixArt-alpha/discussions/31#discussioncomment-7605715) below.

with VAE:
![b7653577-2fbc-4456-9277-34d36fed7b08](https://github.com/PixArt-alpha/PixArt-alpha/assets/48160881/fb795bd4-93f1-4d17-91b7-70d9219e2706)
with TAE:
![25b514ab-79ed-4109-93b5-7294e22650d5](https://github.com/PixArt-alpha/PixArt-alpha/assets/48160881/e84aee11-3e17-4bcc-943f-0ce39310d9a9)

**Comment by lawrence-cj** [comment]
Nice work! The main reference time is spent on model forward, so may I have a detailed time speed up here?

**Reply by muerrilla** [reply]
Sure thing. Although I did a system reboot and now I'm getting better times on both options, than I was getting yesterday, without changing anything. So the speed increase is more like 33% today! 😆

On a 4060ti 16GB, Python 2.1.0+cu11.8 with no CPU offloading and no betterTransformers, Using DPM-Solver for 20 steps at 1024x1024:
Text encoding + sampling takes 10.0 seconds, and decoding takes 2.1 seconds with TAE and 8.2 seconds with VAE. So that's 12.1s vs 18.2s (yesterday I was getting 13.1s vs. 22.2s).

**Reply by lawrence-cj** [reply]
That's a promising result. We test the difference decoder on V100 spec., the result are shown below. I think the inference speed on different GPU spec. really has a big gap.
![B71EB878-AD4F-4A86-B632-6725C2EB165B](https://github.com/PixArt-alpha/PixArt-alpha/assets/55283454/6a53e2e3-0845-4e70-b109-6b1279fddcfb)

**Reply by muerrilla** [reply]
Damn! What are the sampling times like on that beast of a GPU?

**Reply by lawrence-cj** [reply]
The whole process could be done within 8 seconds. 👍 : )

**Reply by muerrilla** [reply]
Damn! I just went through this again and realized the drastic difference was caused by me not having disabled `CUDA - Sysmem Fallback`, so the VAE was pushing the VRAM usage towards the limit and the whole process was being offloaded to RAM (which was not happening when using TAE).

So I fixed it, and now the difference between the two is like 1s. What an embarrassment! 😆🤦‍♂️
